snapcraft: introduce click-based YAML configuration file support

cjp256 opened the pull request:
- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

cjp256 commented:
Requires #2781  and #2780

cjp256 commented:
Rebased.

This PR enables it for remote-build as a limited test case.

I tested it with:
echo -e "launchpad-user: cjp256\nlaunchpad-accept-public-upload: true" > ~/.config/snapcraft/config.yaml

Then running:
`snapcraft remote-build` without args works without prompts.

cjp256 commented:
On a related note, there is an option of utilizing Click's `auto_envvar_prefix` for command groups to enable coverage of environment variable mappings to parameters. https://click.palletsprojects.com/en/7.x/options/#values-from-environment-variables

It would incorporate the command name, e.g. `remote-build` for mapping environment variables, e.g. `SNAPCRAFT_REMOTE_BUILD_LAUNCHPAD_USER=cjp256`.

With the Click wrapper, `_click_load_config()`, we could rely on cmd_name to produce similar behavior if there is some foreseen conflicts among parameter names between commands.  Presumably we would short circuit the lifecycleci commands as the common case.

sergiusens commented on snapcraft/cli/remote.py:
```diff
@@ -22,6 +22,7 @@
 from typing import List
 from xdg import BaseDirectory
 from . import echo
+from .config import enable_snapcraft_config_file
```
why isn't this from ._config ?

sergiusens replied:
It is not meant to be used outside of the cli package, right?

cjp256 replied:
Correct, changed to _config.

cjp256 commented:
Enabled config file for all commands except legacy...

cjp256 commented:
Added info that config file is being used.  Added warning that config file is experimental.

```
$ snapcraft remote-build
Using config file: /home/ubuntu/.config/snapcraft/config.yaml
Configuration file is experimental and is subject to change.
snapcraft remote-build is experimental and is subject to change - use with caution.
```

cjp256 commented:
Rebased

cjp256 commented:
Accidentally forced push wrong branch over this. Fixed.

cjp256 commented:
Scaled it back to just lifecycle and remote commands.

sergiusens commented on snapcraft/cli/lifecycle.py:
```diff
@@ -262,6 +270,7 @@ def snap(directory, output, **kwargs):
 
 
 @lifecyclecli.command(cls=SnapcraftProjectCommand)
+@enable_snapcraft_config_file()
```
pack shouldn't need this

cjp256 replied:
Removed.

cjp256 commented:
rebased/squashed
